tidy3d/components/base.py

```diff
@@ -731,7 +731,49 @@ def __eq__(self, other):
         """Define == for two Tidy3DBaseModels."""
         if other is None:
             return False
-        return self._json_string == other._json_string
+
+        def check_equal(dict1: dict, dict2: dict) -> bool:
+            """Check if two dictionaries are equal, with special handlings."""
+
+            # if different keys, automatically fail
+            if not dict1.keys() == dict2.keys():
+                return False
+
+            # loop through elements in each dict
+            for key in dict1.keys():
+                val1 = dict1[key]
+                val2 = dict2[key]
+
+                # if one of val1 or val2 is None (exclusive OR)
+                if (val1 is None) ^ (val2 is None):
```

Comment:
Might be even better to check that the values have the same type (unless we want to return true if a list has the same items as a tuple, for example): `if type(val1) != type(val2): return False`

This would also eliminate the need for all `or`s in the remaining `if` clauses.

Reply:
one problem with this is that (I think? in ArrayLike) there can be situations where `val1` is a list and `val2` is a `np.ndarray` with the same data. But if we want this situation to be != , then hat you suggest makes a lot of sense.

Reply:
yea this does seem to fail some tests, but might be worth looking into more deeply. Not sure.

Reply:
in this case it was because one was type `float` and one was type `numpy.float64`
<img width="177" alt="image" src="https://github.com/flexcompute/tidy3d/assets/92756888/1e70865a-a915-44cb-925f-45f9198b83f3">

Any ideas on how to handle this type checking maybe more loosely?

Reply:
another thing that comes up is comparing a `tidy3dcomplex` and `complex`.

Reply:
I tried a few different approaches (such as trying to cast val1 to val2's type and vice versa) and it seems there are just a lot of edge cases to handle here, if you have a solution that works, let me know, otherwise I might leave it how it is for now.

Reply:
I see, handling all those cases would be painful. It seems better to leave only the `None` check. We wouldn't be able to foresee user-derived types anyways (a custom extension of `float` for example).